addons/stock/tests/common.py

```diff
@@ -158,25 +178,135 @@ def setUpClass(cls):
+        # TODO: store location and location.id separately
+        cls.stock_location_obj = cls.warehouse_1.lot_stock_id
+        cls.stock_location = cls.warehouse_1.lot_stock_id.id
+        cls.output_location_obj = cls.warehouse_1.wh_output_stock_loc_id
+        cls.output_location = cls.warehouse_1.wh_output_stock_loc_id.id
+        location = cls.warehouse_1.lot_stock_id
+        # location = cls.StockLocationObj.browse(cls.stock_location)
+        if not location.child_ids:
+            cls.StockLocationObj.create([{
+                'name': 'Shelf 1',
+                'location_id': location.id,
+            }, {
+                'name': 'Shelf 2',
+                'location_id': location.id,
+            }])
+
+        # Model Data
+        pack_location = cls.warehouse_1.wh_pack_stock_loc_id
+        pack_location.active = True
+        cls.pack_location = pack_location.id
+        output_location = cls.warehouse_1.wh_output_stock_loc_id
+        output_location.active = True
+        cls.output_location = output_location.id
+        # TODO: fix customer location
+        cls.customer_location = cls.ModelDataObj._xmlid_to_res_id('stock.stock_location_customers')
+        cls.customer_location_obj = cls.env['stock.location'].browse([cls.customer_location])[0]
+        cls.production_location = cls.env['stock.location'].search(
+            [('usage', '=', 'production'), ('company_id', '=', cls.company.id)]
+        )
+        cls.production_location_id = cls.production_location.id
 
         # Existing data
         cls.existing_inventories = cls.env['stock.quant'].search([('inventory_quantity', '!=', 0.0)])
         cls.existing_quants = cls.env['stock.quant'].search([])
 
+    @classmethod
+    def setup_logistics_user(cls):
+        # TODO: is this group setup correct with the new user?
+        cls.env.ref('base.group_user').write({'implied_ids': [
+            (4, cls.env.ref('base.group_multi_company').id),
+            (4, cls.env.ref('stock.group_production_lot').id),
+        ]})
+
+        cls.user = cls.env['res.users'].create({
+            'name': 'Test Stock User',
+            'login': 'stock',
+            'password': 'stockpass',
+            'groups_id': [
+                Command.link(cls.env.ref('stock.group_stock_manager').id),
+            ],
+        })
+        cls.user.email = 'stock@test.com'
+        # Shadow the current environment/cursor with one having the report user.
+        # This is mandatory to test access rights.
+        cls.env = cls.env(user=cls.user)
+        cls.cr = cls.env.cr
+        cls.assign_company_to_user(cls.user, cls.company)
 
-    def url_extract_rec_id_and_model(self, url):
-        rec_id = re.findall(r'[?&]id=([^&]+).*', url)
-        model_name = re.findall(r'[?&]model=([^&]+).*', url)
-        return rec_id, model_name
+    @staticmethod
+    def assign_company_to_user(user, company):
+        # TODO: improve this?
```

Comment:
Seems okay

Reply:
I just don't get the first line

Reply:
The reason for such structure is that, AFAIK, we cannot set the company until it is in the group of allowed companies. So, these 3 lines of code to the following:
- add the company to the group of companies that this user has access to
- set user's company to the one we just added
- remove the remaining companies from the group, leaving our user with access only to the newly added one

Is there perhaps an easier/better way of doing this? :)

Reply:
Ok just add a comment explaining it

Reply:
Done :)